ggml/src/ggml-metal.m

```diff
@@ -1678,33 +1707,49 @@ static void ggml_metal_encode_node(
-                [encoder setBytes:&nb42 length:sizeof(nb42) atIndex:25];
-                [encoder setBytes:&nb50 length:sizeof(nb50) atIndex:26];
-                [encoder setBytes:&nb51 length:sizeof(nb51) atIndex:27];
-                [encoder setBytes:&nb52 length:sizeof(nb52) atIndex:28];
-
-                [encoder dispatchThreadgroups:MTLSizeMake(d_inner, n_seqs, 1) threadsPerThreadgroup:MTLSizeMake(1, 1, 1)];
+                [encoder setBuffer:id_src6 offset:offs_src6 atIndex:6];
+                [encoder setBuffer:id_src7 offset:offs_src7 atIndex:7];
+                [encoder setBuffer:id_dst  offset:offs_dst  atIndex:8];
+
+                [encoder setBytes:&d_state      length:sizeof(d_state)      atIndex:9];
+                [encoder setBytes:&d_inner      length:sizeof(d_inner)      atIndex:10];
+                [encoder setBytes:&n_head       length:sizeof(n_head)       atIndex:11];
+                [encoder setBytes:&n_group      length:sizeof(n_group)      atIndex:12];
+                [encoder setBytes:&n_seq_tokens length:sizeof(n_seq_tokens) atIndex:13];
+                [encoder setBytes:&n_seqs       length:sizeof(n_seqs)       atIndex:14];
+
+                [encoder setBytes:&nb00 length:sizeof(nb00) atIndex:15];
+                [encoder setBytes:&nb01 length:sizeof(nb01) atIndex:16];
+                [encoder setBytes:&nb02 length:sizeof(nb02) atIndex:17];
+                [encoder setBytes:&nb03 length:sizeof(nb03) atIndex:18];
+                [encoder setBytes:&nb10 length:sizeof(nb10) atIndex:19];
+                [encoder setBytes:&nb11 length:sizeof(nb11) atIndex:20];
+                [encoder setBytes:&nb12 length:sizeof(nb12) atIndex:21];
+                [encoder setBytes:&nb13 length:sizeof(nb13) atIndex:22];
+                [encoder setBytes:&nb20 length:sizeof(nb20) atIndex:23];
+                [encoder setBytes:&nb21 length:sizeof(nb21) atIndex:24];
+                [encoder setBytes:&nb22 length:sizeof(nb22) atIndex:25];
+                [encoder setBytes:&nb23 length:sizeof(nb23) atIndex:26];
+                [encoder setBytes:&nb30 length:sizeof(nb30) atIndex:27];
+                [encoder setBytes:&nb31 length:sizeof(nb31) atIndex:28];
+                [encoder setBytes:&nb40 length:sizeof(nb40) atIndex:29];
+                [encoder setBytes:&nb41 length:sizeof(nb41) atIndex:30];
+                [encoder setBytes:&nb42 length:sizeof(nb42) atIndex:31];
+                [encoder setBytes:&nb43 length:sizeof(nb43) atIndex:32];
+                [encoder setBytes:&nb50 length:sizeof(nb50) atIndex:33];
+                [encoder setBytes:&nb51 length:sizeof(nb51) atIndex:34];
+                [encoder setBytes:&nb52 length:sizeof(nb52) atIndex:35];
+                [encoder setBytes:&nb53 length:sizeof(nb53) atIndex:36];
+                [encoder setBytes:&nb60 length:sizeof(nb60) atIndex:37];
+                [encoder setBytes:&nb70 length:sizeof(nb70) atIndex:38];
```

Comment:
Does this work? I think the limit for encoder arguments is 32.

Reply:
I remember having some sort of issues with using too much `setBytes` for a kernel. I think the relevant limit is this one: 

![image](https://github.com/user-attachments/assets/259f96a1-f923-48a5-9efd-cd5d8e4bb3cb)

https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf

Just tried building on M2 Ultra and I get these errors:

```bash
make -j tests && ./tests/test-backend-ops -b Metal -o SSM_SCAN

...
Testing 2 backends

Backend 1/2 (CPU)
  Skipping
Backend 2/2 (Metal)
ggml_metal_init: allocating
ggml_metal_init: found device: Apple M2 Ultra
ggml_metal_init: picking default device: Apple M2 Ultra
ggml_metal_init: using embedded metal library
ggml_metal_init: error: Error Domain=MTLLibraryErrorDomain Code=3 "program_source:2632:19: warning: unused variable 'ncs' [-Wunused-variable]
    const int64_t ncs = ne00;
                  ^
program_source:2633:19: warning: unused variable 'nr' [-Wunused-variable]
    const int64_t nr  = ne01;
                  ^
program_source:2634:19: warning: unused variable 'n_t' [-Wunused-variable]
    const int64_t n_t = ne1;
                  ^
program_source:2635:19: warning: unused variable 'n_s' [-Wunused-variable]
    const int64_t n_s = ne2;
                  ^
program_source:2684:29: error: no 'buffer' resource location available for 'nb43'
        constant uint64_t & nb43,
                            ^
program_source:2685:29: error: no 'buffer' resource location available for 'nb50'
        constant uint64_t & nb50,
                            ^
program_source:2686:29: error: no 'buffer' resource location available for 'nb51'
        constant uint64_t & nb51,
                            ^
program_source:2687:29: error: no 'buffer' resource location available for 'nb52'
        constant uint64_t & nb52,
                            ^
program_source:2688:29: error: no 'buffer' resource location available for 'nb53'
        constant uint64_t & nb53,
                            ^
program_source:2689:29: error: no 'buffer' resource location available for 'nb60'
        constant uint64_t & nb60,
                            ^
program_source:2690:29: error: no 'buffer' resource location available for 'nb70'
        constant uint64_t & nb70,
...
```